pennylane/templates/base/single.py

```diff
@@ -0,0 +1,240 @@
+                return qml.expval(qml.PauliZ(0))
+
+            circuit([[1, 1], [2, 1], [0.1, 1]])
+
+        As mentioned above, in general the unitary **must** have the following signature:
+
+        .. code-block:: python
+
+            unitary(parameter1, parameter2, ... parameterD, wires, **kwargs)
+
+        If ``unitary`` does not depend on parameters (:math:`D=0`), the signature is
+
+        .. code-block:: python
+
+            unitary(wires, **kwargs)
+
+        As a result, ``parameters`` must be a list or array of length-:math:`D` lists or arrays.
+
+        If :math:`D` becomes large, the signature can be simplified by wrapping each entry in ``parameters``:
+
+        .. code-block:: python
+
+            @template
+            def mytemplate(pars, wires):
+                qml.Hadamard(wires=wires)
+                qml.RY(pars[0], wires=wires)
+                qml.RX(pars[1], wires=wires)
+
+            @qml.qnode(dev)
+            def circuit(pars):
+                Single(unitary=mytemplate, wires=[0,1,2], parameters=pars)
+                return qml.expval(qml.PauliZ(0))
+
+            print(circuit([[[1, 1]], [[2, 1]], [[0.1, 1]]]))
+
+        If the number of parameters for each wire does not match the template or gate, an error gets thrown:
+
+        .. code-block:: python
+
+                @template
+                def mytemplate(pars1, pars2, wires):
+                    qml.Hadamard(wires=wires)
+                    qml.RY(pars1, wires=wires)
+                    qml.RX(pars2, wires=wires)
+
+                @qml.qnode(dev)
+                def circuit(pars):
+                    Single(unitary=mytemplate, wires=[0, 1, 2], parameters=pars)
+                    return qml.expval(qml.PauliZ(0))
+
+
+        >>> circuit([1, 2, 3]))
+        TypeError: mytemplate() missing 1 required positional argument: 'pars2'
+
+        **Keyword arguments**
+
+        The unitary can be a template that takes additional keyword arguments.
+
+        .. code-block:: python
+
+            @template
+            def mytemplate(wires, h=True):
+                if h:
+                    qml.Hadamard(wires=wires)
+                qml.T(wires=wires)
+
+            @qml.qnode(dev)
+            def circuit(hadamard=None):
+                Single(unitary=mytemplate, wires=[0, 1, 2], kwargs={'h': hadamard})
+                return qml.expval(qml.PauliZ(0))
+
+            circuit(hadamard=False)
+
+    """
+
+    #########
+    # Input checks
+
+    wires = _check_wires(wires)
+
+    _check_type(parameters, [Iterable, type(None)], msg="'parameters' must be either of type None or "
+                                                        "Iterable; got {}".format(type(parameters)))
+#    _check_type(unitary, [list, FunctionType], msg="unitary must be a ``~.pennylane.operation.Operation`` "
+#                                                   "or a template function; got {}".format(type(unitary)))
```

Comment:
How come this check is commented out?

Reply:
Oh, thanks, that was due to a problem related to the fact that gates, when evaluated, are of type ``Operation``, but else they are functions. Will fix...